File: observer-pattern/weather-station.py
# Subject
class WeatherStation:
  def __init__(self):
    self._observers = []
    self._temperature = 0

  def add_observer(self, observer):
    self._observers.append(observer)

  def remove_observer(self, observer):
    self._observers.remove(observer)

  def notify_observers(self):
    """Notify all observers about the temperature change"""
    for obs in self._observers:
      obs.update(self._temperature)

  def set_temperature(self, temp):
    """Set the temperature and notify observers"""
    print(f'Setting temperature to {temp} degrees')
    self._temperature = temp
    self.notify_observers()


# Observer (interface)
class Observer:
  def update(self, temp):
    """React to temperature updates"""
    raise NotImplementedError

# Concrete observer: Say a display device
class DisplayDevice(Observer):
  def update(self, temperature):
    print(f'Display device (observer): Temperature updated to {temperature} degrees')
    # We can trigger some code here to update the display value according to the value of temperature received.
    # print('Display pixels updated')


# Dry run
if __name__ == '__main__':
  station = WeatherStation()
  display_device = DisplayDevice()

  station.add_observer(display_device) # Register display device as an observer
  
  import time
  import random

  for i in range(3):
    sleep_time = random.randint(1, 3) * i

    station.set_temperature(14 * i + 3)

    print(f'[DEBUG]: Sleeping for {sleep_time} seconds \n\n')
    time.sleep(sleep_time)

File: observer-pattern/weather-station/weather-station.rs
use std::cell::RefCell;
use std::rc::Rc;

// Abstract Observer
trait Observer {
    fn update(&self, temperature: u32);
}

// Concrete Observer
struct DisplayDevice;

impl Observer for DisplayDevice {
    fn update(&self, temperature: u32) {
        println!(
            "Display Device: Temperature updated to {} degrees.",
            temperature
        );
    }
}

// Subject
struct WeatherStation {
    temperature: u32,
    observers: Vec<Rc<RefCell<dyn Observer>>>,
}

impl WeatherStation {
    fn new() -> Self {
        Self {
            temperature: 0,
            observers: Vec::new(),
        }
    }

    fn add_observer(&mut self, observer: Rc<RefCell<dyn Observer>>) {
        self.observers.push(observer);
    }

    fn remove_observer(&mut self, observer: Rc<RefCell<dyn Observer>>) {
        self.observers.retain(|obs| !Rc::ptr_eq(obs, &observer));
    }

    fn notify_observers(&self) {
        for obs in &self.observers {
            obs.borrow().update(self.temperature);
        }
    }

    fn set_temperature(&mut self, temperature: u32) {
        println!(
            "Weather Station: Before updating temperature to {} degrees.",
            temperature
        );
        self.temperature = temperature;
        self.notify_observers();
    }
}

fn main() {
    let mut station = WeatherStation::new();
    let display_device = Rc::new(RefCell::new(DisplayDevice));

    station.add_observer(display_device.clone());

    station.set_temperature(25);

    station.set_temperature(32);

    station.remove_observer(display_device.clone());

    println!("Removed display device as an observer, updating temp again.");
    station.set_temperature(28);

    println!("Adding display device as an observer again.");
    station.add_observer(display_device.clone());

    station.set_temperature(30);
}
